lib/type_check/builtin/fixed_map.ex

```diff
@@ -40,11 +40,11 @@ defmodule TypeCheck.Builtin.FixedMap do
 
     defp map_check(param, s) do
       quote generated: true, location: :keep do
-        x = unquote(param)
-        if is_map(x) do
-          {:ok, [], x}
-        else
-          {:error, {unquote(Macro.escape(s)), :not_a_map, %{}, x}}
+        case unquote(param) do
+          val when is_map(val) ->
+            {:ok, [], val}
+            other ->
+            {:error, {unquote(Macro.escape(s)), :not_a_map, %{}, other}}
```

Comment:
I am really confused.
I assumed `unquote(param)` is evaluating directly to a map (not to a variable that can be a map but to a data structure like `%{}`).

In that scenario:

```
    if is_map(%{}) do
      IO.puts "a"
    else
      IO.puts "b"
    end
```
produces:
```
The pattern can never match the type.

Pattern:
false

Type:
true
```
but then
```
    case %{} do
      val when is_map(val) -> :ok
      other -> other
    end
```
should produce a different error:
```
The pattern
variable_other

can never match, because previous clauses completely cover the type
%{}.
```

But I tried this branch and the fix works. I wonder, how were you able pinpoint that this is the exact point that dialyzer complains about? I'd like to try myself and see what `unquote(param)` evaulets to that it throws error on `if` statement but equivalent `case` statement is somehow fine :)

Reply:
Oh! You are right with suspecting Elixir's `if`.
```
quoted = quote do if true, do: :a, else: :b end
quoted |> Macro.expand(__ENV__) |> Macro.to_string |> IO.puts

case(true) do
  x when Kernel.in(x, [false, nil]) ->
    :b
  _ ->
    :a
end
```

Dialyzer is able to track the types in `Kernel.in` and since `true` cannot ever match `false` (or `nil` but `false` is first), it decides that this is a redundant clause. The funny thing is, that `if false` does not produce the same error because it matches the first clause but somehow dialyzer can't figure out the second clause is redundant (probably because of `Kernel.in` usage).

But I still can't figure out how rewriting an `if` to a `case` helped :shrug:

Reply:
I used `use TypeCheck, debug: true` to look at all Elixir code that is generated for a particular spec helper.
By placing this in a separate module in a compiled file (and renaming the module name to allow it to exist alongside with the original module) it is possible to see in more details what particular line(s) of the generated spec code Dialyzer was complaining about.

I then tried fixing it by changing what code is generated, and then running `mix compile` again before re-running Dialyzer.

---

What is a bit disheartening is that there are many other situations Dialyzer complains about, but which are 'fine', i.e. there are situations in which someone indeed calls a function with a hard-coded valid input and this results in the error-parts of the spec code never being called. In some cases the Erlang compiler is even able to optimize it away.
However, Dialyzer will complain and state that we've written superfluous code.

I have no clue at this time how to improve this situation. Maybe there is a way to communicate that superfluous branches are fine in a particular module to Dialyzer?
Help with this is very much appreciated, as I am a bit out of my depth here.

Reply:
I don't know about any way to silence dialyzer. The only thing that comes to my mind is to perform branching logic outside of `quote`. E.g. instead of:
```
quote do
  if is_map(unquote(x)) do
    :a
  else
    :b:
  end
end
```
something like:
```
if is_map(x) do
  quote do
    :a
  end
else
  quote do
    :b:
  end
end
```
That would eliminate conditions from generated code by moving branching logic to compilation time.
However, I am not sure how feasible that idea is.